File: legacy/src/main/rs/emulate/legacy/map/MapPlane.java
package rs.emulate.legacy.map;

/**
 * A plane of a map, which is a distinct height level.
 *
 * @author Major
 */
public final class MapPlane {

	/**
	 * The 2-dimensional array of Tiles.
	 */
	private final Tile[][] tiles;

	/**
	 * The level of this MapPlane.
	 */
	private final int level;

	/**
	 * Creates the MapPlane.
	 *
	 * @param level The level of the MapPlane.
	 * @param tiles The 2-dimensional array of {@link Tile}s.
	 */
	public MapPlane(int level, Tile[][] tiles) {
		this.level = level;
		this.tiles = clone(tiles);
	}

	/**
	 * Gets the {@link Tile} at the specified (x, z) coordinate.
	 *
	 * @param x The x coordinate.
	 * @param z The z coordinate.
	 * @return The Tile.
	 */
	public Tile get(int x, int z) {
		return tiles[x][z];
	}

	/**
	 * Gets the level of this MapPlane.
	 *
	 * @return The level.
	 */
	public int getLevel() {
		return level;
	}

	/**
	 * Returns a shallow copy of the specified 2-dimensional array.
	 *
	 * @param array The array to copy.
	 * @return The copy.
	 */
	private static <T> T[][] clone(T[][] array) {
		T[][] copy = array.clone();
		for (int index = 0; index < copy.length; index++) {
			copy[index] = array[index].clone();
		}

		return copy;
	}

}

File: common/src/main/kotlin/rs/emulate/common/map/MapPlane.kt
package rs.emulate.common.map

/**
 * A plane of a map, which is a distinct height level.
 *
 * @param level The height level of the MapPlane.
 * @param tiles The 2D array of [Tile]s. Must be square.
 */
class MapPlane(val level: Int, tiles: Array<Array<Tile>>) {

    private val tiles: Array<Array<Tile>> = clone(tiles)

    /**
     * The amount of tiles in this MapPlane.
     */
    val size: Int = tiles.size * tiles.size

    /**
     * Returns a shallow copy of the specified 2-dimensional array.
     */
    private fun <T> clone(array: Array<Array<T>>): Array<Array<T>> {
        val copy = array.clone()
        for (index in copy.indices) {
            copy[index] = array[index].clone()
        }

        return copy
    }

    /**
     * Gets the [Tile] at the specified (x, z) coordinate.
     */
    fun getTile(x: Int, z: Int): Tile {
        return tiles[x][z]
    }

    /**
     * Gets the [Tile]s in this MapPlane.
     *
     * This method returns the Tiles according on a column-based ordering: for a 2x2 tile set, the order will be
     * `(0, 0), (0, 1), (1, 0), (1, 1)`.
     */
    fun tiles(): List<Tile> {
        return tiles.flatMap { it.toList() }
    }

}
